docs/docs/tutorials/adding_annotations.md

```diff
@@ -0,0 +1,184 @@
+
+## Generating the grpc-gateway stubs
+
+Now that we've got the grpc-gateway annotations added to the proto file, we need to use the grpc-gateway generator to generate the stubs.
+
+Before we can do that, we need to copy some dependencies into our protofile structure. Copy the `third_party/googleapis` folder from the grpc-gateway repository to your local protofile structure. It should look like this afterwards:
+
+```
+proto
+├── google
+│   └── api
+│       ├── annotations.proto
+│       └── http.proto
+└── helloworld
+    └── hello_world.proto
+```
+
+### Using buf
+
+We'll need to add the grpc-gateway generator to the generation configuration:
+
+```yml
+version: v1beta1
+plugins:
+  - name: go
+    out: proto
+    opt: paths=source_relative
+  - name: go-grpc
+    out: proto
+    opt: paths=source_relative,require_unimplemented_servers=false
+  - name: grpc-gateway
+    out: proto
+    opt: paths=source_relative
+```
+
+And that's it! Now if you run:
+
+```sh
+$ buf generate
+```
+
+It should produce a `*.gw.pb.go` file.
+
+### Using protoc
+
+Now we need to add the grpc-gateway generator to the protoc invocation:
+
+```sh
+$ protoc -I ./proto \
+  --go_out ./proto --go_opt paths=source_relative \
+  --go-grpc_out ./proto --go-grpc_opt paths=source_relative \
+  --grpc-gateway_out ./proto --grpc-gateway_opt paths=source_relative \
+  ./proto/helloworld/hello_world.proto
+```
+
+This should generate a `*.gw.pb.go` file.
+
+Usage examples can be found on this [Usage](https://github.com/grpc-ecosystem/grpc-gateway#usage)
+
+### In addition to the main.go
+
+```go
+package gateway
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	helloworldpb "github.com/iamrajiv/helloworld/proto/helloworld"
+	"github.com/prometheus/common/log"
+	"google.golang.org/grpc"
+)
+
+// Run runs the gRPC-Gateway, dialling the provided address.
+func Run(dialAddr string) error {
+	// Create a client connection to the gRPC Server we just started.
+	// This is where the gRPC-Gateway proxies the requests.
+	conn, err := grpc.DialContext(
+		context.Background(),
+		dialAddr,
+		grpc.WithBlock(),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to dial server: %w", err)
+	}
+
+	gwmux := runtime.NewServeMux()
+	err = helloworldpb.RegisterGreeterHandler(context.Background(), gwmux, conn)
+	if err != nil {
+		return fmt.Errorf("failed to register gateway: %w", err)
+	}
+
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "11000"
+	}
+	gatewayAddr := "0.0.0.0:" + port
+	gwServer := &http.Server{
+		Addr: gatewayAddr,
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if strings.HasPrefix(r.URL.Path, "") {
+				gwmux.ServeHTTP(w, r)
+				return
+			}
+		}),
+	}
+	log.Info("Serving gRPC-Gateway and OpenAPI Documentation on http://", gatewayAddr)
+	return fmt.Errorf("serving gRPC-Gateway server: %w", gwServer.ListenAndServe())
+}
+```
```

Comment:
I was hoping we could just merge the existing main.go with the gateway here, not just paste the "Run" method. That's not very useful to the user. I've got a suggestion here, can you apply this and try to run it yourself?
```suggestion
package main

import (
	"context"
	"io/ioutil"
	"net"
	"os"
	"log"

	"google.golang.org/grpc"
	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"

	helloworldpb "github.com/myuser/myrepo/proto/helloworld"
)

type server struct{}

func NewServer() *server {
	return &server{}
}

func (s *server) SayHello(ctx context.Context, in *helloworldpb.HelloRequest) (*helloworldpb.HelloReply, error) {
	return &helloworldpb.HelloReply{Message: in.Name + " World"}, nil
}

func main() {
	lis, err := net.Listen("tcp", ":8080")
	if err != nil {
		log.Fatalln("Failed to listen:", err)
	}
	s := grpc.NewServer()
	helloworldpb.RegisterGreeterServer(s, &server{})
	// Serve gRPC Server
	log.Println("Serving gRPC on 0.0.0.0:8080")
	go func() {
		log.Fatalln(s.Serve(lis))
	}()

	// Create a client connection to the gRPC Server we just started.
	// This is where the gRPC-Gateway proxies the requests.
	conn, err := grpc.DialContext(
		context.Background(),
		"0.0.0.0:8080",
		grpc.WithBlock(),
	)
	if err != nil {
		log.Fatalln("Failed to dial server:", err)
	}

	gwmux := runtime.NewServeMux()
	err = helloworldpb.RegisterGreeterHandler(context.Background(), gwmux, conn)
	if err != nil {
		log.Fatalln("Failed to register gateway:", err)
	}
	gwServer := &http.Server{
		Addr: ":8090",
		Handler: gwmux,
	}
	log.Println("Serving gRPC-Gateway on http://0.0.0.0:8090")
	log.Fatalln(gwServer.ListenAndServe())
}
''' (backticks)
```